File: codediffparser/Python/codediffparser/codediffparser/node.py
from .qualified_name_helper import QualifiedNameHelper
from .status import Status
from .node_type import NodeType


class Node():

    def __init__(self):
        self.file_path = ""
        self.name = ""
        self.declaring_classes_name = ""
        self.project_name = "" # == packageName
        self.type = NodeType.UNKNOWNFILE
        #self.modifiers = []
        self.position = [] # target: starting line, ending line
        self.position_old = [] # source: starting line, ending line
        self.status = Status.UNCHANGED
        self.parent_node_id = None
        self.param_hash_code = "0"
        self.node_hash_code = "0"
        self.is_generated = False
        self.language = "Python"
        self.notes = "" # Is used for error messages.
    
    
    def get_id(self):
        if self.type is NodeType.METHOD or self.type is NodeType.METHOD_REFERENCE:
            return QualifiedNameHelper.concat_save_with_delimiter("#", self.parent_node_id, self.name) + str(self.param_hash_code)
        elif self.type is NodeType.FUNCTION or self.type is NodeType.FUNCTION_REFERENCE:
            return QualifiedNameHelper.concat_save(self.project_name, self.declaring_classes_name, self.name) + str(self.param_hash_code)
        elif self.type is NodeType.SCRIPT:
            return self.project_name
        elif self.type is NodeType.UNKNOWNFILE:
            return self.file_path
        elif self.type is NodeType.TOOLERROR:
            return str(self.node_hash_code)
        else:
            return QualifiedNameHelper.concat_save(self.project_name, self.declaring_classes_name, self.name)

    
    def set_position(self, start_line, end_line):
        self.position = [start_line, end_line]

    def set_position_old(self, start_line, end_line):
        self.position_old = [start_line, end_line]
    

    def json_helper(self):
        """Export node to json."""
        obj = {}
        obj['filePath'] = self.file_path
        obj['name'] = self.name
        obj['declaringScopesName'] = self.declaring_classes_name
        obj['packageName'] = self.project_name
        obj['type'] = self.type.name
        #obj['modifiers'] = self.modifiers
        obj['position'] = self.position
        obj['positionOld'] = self.position_old
        obj['status'] = self.status.name
        obj['parentNodeId'] = self.parent_node_id
        obj['id'] = self.get_id()
        obj['paramHashCode'] = self.param_hash_code
        obj['nodeHashCode'] = self.node_hash_code
        obj['generated'] = self.is_generated
        obj['language'] = "" if self.type is NodeType.UNKNOWNFILE else "Python"
        obj['notes'] = self.notes
        return obj

File: codediffparser/Java/codediffparser/src/main/java/codediffparser/callgraph/Node.java
package codediffparser.callgraph;

import static codediffparser.utils.QualifiedNameHelper.concatSave;
import static codediffparser.utils.QualifiedNameHelper.concatSaveWithDelimiter;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Node {
    private String filePath;
    private String name;
    private String declaringScopesName;
    private String packageName;
    private NodeType type;
    private List<String> modifiers;
    private Integer[] position;
    private Integer[] positionOld;
    private Status status;
    private String parentNodeId;
    private int paramHashCode;
    private int nodeHashCode;
    private boolean isGenerated;
    private String language;

    private Node() {
    }

    public enum NodeType {
        CLASS,
        INTERFACE,
        ABSTRACT_CLASS,
        METHOD,
        TYPE_REFERENCE,
        METHOD_REFERENCE,
        UNKNOWNFILE;

        public boolean isDeclaredType() {
            if (List.of(CLASS, INTERFACE, ABSTRACT_CLASS, METHOD).contains(this)) {
                return true;
            }
            return false;
        }

        public boolean isReferencedType() {
            if (List.of(TYPE_REFERENCE, METHOD_REFERENCE).contains(this)) {
                return true;
            }
            return false;
        }
    }

    public String getId() {
        if (type == NodeType.METHOD || type == NodeType.METHOD_REFERENCE) {
            return concatSave(packageName, concatSaveWithDelimiter("#", declaringScopesName, name)) + paramHashCode;
        } else if (type == NodeType.UNKNOWNFILE) {
            return filePath;

        } else {
            return concatSave(packageName, declaringScopesName, name);
        }
    }

    public String getFilePath() {
        return filePath;
    }

    public String getName() {
        return name;
    }

    public String getDeclaringScopesName() {
        return declaringScopesName;
    }

    public String getPackageName() {
        return packageName;
    }

    public NodeType getType() {
        return type;
    }

    public void setType(NodeType type) {
        this.type = type;
    }

    public List<String> getModifiers() {
        if (modifiers == null) {
            modifiers = new ArrayList<>();
        }
        return modifiers;
    }

    public Integer[] getPosition() {
        return position;
    }

    public Integer[] getPositionOld() {
        return positionOld;
    }

    public void setPositionOld(Integer[] positionOld) {
        if (position == null) {
            this.position = positionOld;
        } else if (!Arrays.equals(position, positionOld)) {
            this.positionOld = positionOld;
        }
    }

    public Status getStatus() {
        return status;
    }

    public String getParentNodeId() {
        return parentNodeId;
    }

    public void addModifier(String modifier) {
        getModifiers().add(modifier);
    }

    public void setStatus(Status status) {
        this.status = status;
    }

    public int getNodeHashCode() {
        return nodeHashCode;
    }

    public boolean isGenerated() {
        return isGenerated;
    }

    public void setGenerated(boolean isGenerated) {
        this.isGenerated = isGenerated;
    }

    public static class Builder extends AbstractBuilder<Node> {
        private String filePath;
        private String name;
        private String declaringScopesName;
        private String packageName;
        private NodeType type;
        private List<String> modifiers;
        private Integer[] position;
        private String parentNodeId;
        private int paramHashCode;
        private int nodeHashCode;
        private boolean isGenerated = false;

        public Builder withFilePath(String filePath) {
            this.filePath = filePath;
            return this;
        }

        public Builder withName(String name) {
            this.name = name;
            return this;
        }

        public Builder withDeclaringScopesName(String declaringScopesName) {
            this.declaringScopesName = declaringScopesName;
            return this;
        }

        public Builder withPackageName(String packageName) {
            this.packageName = packageName;
            return this;
        }

        public Builder withType(NodeType type) {
            this.type = type;
            return this;
        }

        public Builder withModifiers(List<String> modifiers) {
            this.modifiers = modifiers;
            return this;
        }

        public Builder withPosition(Integer lineNumberStart, Integer lineNumberEnd) {
            this.position = new Integer[] { lineNumberStart, lineNumberEnd };
            return this;
        }

        public Builder withParentNodeId(String parentNodeId) {
            this.parentNodeId = parentNodeId;
            return this;
        }

        public Builder withParamHashCode(int paramHashCode) {
            this.paramHashCode = paramHashCode;
            return this;
        }

        public Builder withNodeHashCode(int nodeHashCode) {
            this.nodeHashCode = nodeHashCode;
            return this;
        }

        public Builder withIsGenerated(boolean isGenerated) {
            this.isGenerated = isGenerated;
            return this;
        }

        @Override
        public Node build() {
            final Node node = new Node();
            node.filePath = this.filePath;
            node.name = this.name;
            node.declaringScopesName = this.declaringScopesName;
            node.packageName = this.packageName;
            node.type = this.type;
            node.modifiers = this.modifiers;
            node.position = this.position;
            node.parentNodeId = this.parentNodeId;
            node.paramHashCode = this.paramHashCode;
            node.nodeHashCode = this.nodeHashCode;
            node.isGenerated = this.isGenerated;
            node.language = this.type == NodeType.UNKNOWNFILE ? "" : "Java";

            nullCheck(name, type);
            return node;
        }
    }
}
